src/connector/src/sink/kafka.rs

```diff
@@ -439,15 +452,19 @@ impl KafkaPayloadWriter {
                 Err((e, rec)) => {
                     record = rec;
                     match e {
-                        err @ KafkaError::MessageProduction(RDKafkaErrorCode::QueueFull)
-                        | err @ KafkaError::MessageProduction(RDKafkaErrorCode::MessageTimedOut) => {
+                        err @ KafkaError::MessageProduction(RDKafkaErrorCode::QueueFull) => {
                             tracing::warn!(
                                 "producing message (key {:?}) to topic {} failed, err {:?}, retrying",
                                 record.key.map(|k| k.to_bytes()),
                                 record.topic,
                                 err
                             );
-                            tokio::time::sleep(self.config.retry_interval).await;
+                            Self::map_future_result(
```

Comment:
IIUC, here await here is to detect other potential errors like `MessageTimeOut` when the current buffer queue in `rdkafka` is full? Then I think it's reasonable to await until `max_delivery_buffer_size` just like the `Ok()` case.

Reply:
We looked at the code of rdkafka yesterday, and `MessageTimeOut` is unlikely to be returned in the `send_result`. The general logic here is to leverage the internal queue of kafka client to limit the sent but not delivered messages, and when we detect a `QueueFull`, we want for any future delivery to release the quota in the queue. In this way the queue size can be implicitly aligned with the internal kafka client setting. The `max_delivery_buffer_size=queue_size*1.2` is a backup logic and it's not likely to reach there if the kafka client queue logic is as expected.

Reply:
Make sense! Thanks for explanation🥰